### PR TITLE
Shaka Streamer Binaries

### PR DESCRIPTION
**Pull request by meryacine**
## Steps to build:
- run `python build_wheels.py` in the terminal.

### CODE REVIEW

File: binaries/build.py
Comment (joeyparrish) [inline binaries/build.py:46]: If you have time, it would be cool if build.py had constants for the release tags of Packager & ffmpeg, and would download the binaries from the GitHub repos automatically, rather than requiring us to download them manually and in advance of build.py.

File: streamer/controller_node.py
Comment (joeyparrish) [inline streamer/controller_node.py:104]: Please remove the "run as root" recommendation.  If it doesn't work at runtime, we need to solve it at packaging time instead, and I'll commit to doing that.  We definitely shouldn't run all of streamer as root, and I don't want to recommend questionable security practices.

If we ultimately can't solve it at packaging time, I'll craft a message here that recommends a specific, targeted command to run as root/admin to fix the permissions.

File: streamer/controller_node.py
Comment (joeyparrish) [inline streamer/controller_node.py:116]: typo: binaries

File: streamer/controller_node.py
Comment (joeyparrish) [inline streamer/controller_node.py:307]: typo: packager

Though, no joke, as I tried to type "typo" in that comment, I first accidentally typed "topy".  😄

**Review by joeyparrish — CHANGES_REQUESTED** [review]
This looks great!  Just a few comments.

File: binaries/build.py
Comment (meryacine) [reply binaries/build.py:46]: Sure :)

File: binaries/download_binaries.py
Comment (meryacine) [inline binaries/download_binaries.py:140]: @joeyparrish As you see, it is not so reliable way to produce the yaml file we use for building the wheels xD.
Would you please name the new upcoming ARM static build with a name that has `arm64` in it, or maybe `aarch64`, and have an `x64` for linux then so we can differentiate between the two linux es.
We can also do the same for windows and macos, just for clarity.
`packager-win-x64.exe`, `packager-osx-x64.dmg` something like this.

**Review by michellezhuogg — APPROVED** [review]

**Review by michellezhuogg — APPROVED** [review]

File: binaries/download_binaries.py
Comment (joeyparrish) [reply binaries/download_binaries.py:140]: Totally.  I will update Packager's workflows to match the naming scheme I used for FFmpeg.

File: binaries/build.py
Comment (joeyparrish) [inline binaries/build.py:39]: I don't understand the use of `__name__` here.  Isn't `streamer_binaries.__name__` the same as `"streamer_binaries"`?

File: binaries/package_data.yaml
Comment (joeyparrish) [inline binaries/package_data.yaml:5]: This file looks incomplete (no packager binary names for linux) and has no license header.  Is it generated?  Should this be in .gitignore?

File: binaries/platform_package_data.yaml
Comment (joeyparrish) [inline binaries/platform_package_data.yaml:29]: Could you add a TODO after this to clean up the Packager release file names?

File: binaries/download_binaries.py
Comment (joeyparrish) [inline binaries/download_binaries.py:152]: Best practice is to have this inside a conditional:

```py
if __name__ == '__main__':
  main()
```

Then main() will not run as a side-effect of an import.  Another script or module loading this one will have to call main() explicitly, which is meant to cut down on surprises and accidents.

File: binaries/download_binaries.py
Comment (joeyparrish) [inline binaries/download_binaries.py:148]: Is platform_package_data.yaml still needed in the repo?  Should both yaml files be removed from the PR, and the generated one added to .gitignore?

File: binaries/download_binaries.py
Comment (joeyparrish) [inline binaries/download_binaries.py:149]: Since you have modules importing modules, why pass data in a yaml file instead of a dictionary return value?

File: binaries/download_binaries.py
Comment (joeyparrish) [inline binaries/download_binaries.py:92]: It seems like overkill to use threads to download binaries.  I would prefer something simpler, even if it's somewhat slower.

File: binaries/download_binaries.py
Comment (joeyparrish) [inline binaries/download_binaries.py:122]: I think versioned download URLs could be constructed without resorting to the GitHub API.

Something like:

```py
PACKAGER_VERSION = 'v2.5.1'
FFMPEG_VERSION = 'early-trial-122`  # FIXME: update when first official release made

PACKAGER_DL_PREFIX = 'https://github.com/google/shaka-packager/releases/download/' + PACKAGER_VERSION
PACKAGER_DL = {
  'win-x64': PACKAGER_DL_PREFIX + '/packager-win.exe',
  'osx-x64': PACKAGER_DL_PREFIX + '/packager-osx',
}

# etc...
```

I think the download script could be much simpler this way.  Updates would just involve updating the tags at the top.

File: binaries/build.py
Comment (joeyparrish) [inline binaries/build.py:95]: Could you consider using [`TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) instead?  Something like:

```py
with tempfile.TemporaryDirectory() as temp_dir_name:
  download_into(temp_dir_name, platform_name)
  build_binary_package(temp_dir_name, platform_name)
# Now the temporary directory and its contents have been wiped out
```

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: binaries/download_binaries.py
Comment (meryacine) [reply binaries/download_binaries.py:149]: Sure,
I also think of combining the downloader and the builder files together.

File: binaries/build.py
Comment (meryacine) [reply binaries/build.py:39]: Thought this just might be a better way to refer to the package's name. Yes, they are the same.

File: binaries/build.py
Comment (meryacine) [reply binaries/build.py:95]: If i get this correctly, you mean to download the binaries in a temp directory and then add it to the package.
`setup` looks for `package_data` inside the package we say it resides in(in this case, inside `streamer_binaries`), i don't think we can put it some where else. If `setup` couldn't find the files we listed in `package_data` inside the package, it just skips it, doesn't search somewhere else.

Comment (meryacine) [reply binaries/build.py]: I looked up a bit about whether we are able to specify the build directory to the `bdist_wheel` command, but found we no way.  `bdist_wheel` calls `build` internally, while the `build` command has the `--build-lib` option exposed, the `bdist_wheel` doesn't.
Maybe the usage of `cmdclass` is a bit confusing.  We can simplify this to be only creating the needed build directory before the `setup` call and removing it afterwards, after the `setup` call.

Comment (meryacine) [reply binaries/build.py]: I did change it a little bit now to use the temporary directory `--bdist-dir`. This means we build directly in the same directory we package things in afterwards, `setuptools` removes this directory after building the wheel.
Kinda hacky, but i think is simpler.

**Comment by joeyparrish** [comment]
Just back from vacation.  Will try to take another pass at review today if at all possible.

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:30]: Not a real version.  :-)  v2.5.1 is out, and v2.6.0 is coming soon (< 1 week, I think).  The consistent binary suffixes are going to be in v2.6.0, so you can go ahead and pin to that version.

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:47]: Please change to google/shaka-packager.  I test some CI changes on my fork, but that is not where we will officially take binaries from.

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:55]: style nit: align with `BINARIES_DL`

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:44]: style nit: align with `PLATFORM_POSTFIXES`

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:35]: Perhaps rename to "suffix" instead of "postfix".  I checked myself on Wikipedia, since perhaps I'm just not used to hearing postfix for the ending of a word.  What I found was the [disambiguation article](https://en.wikipedia.org/wiki/Postfix) that says:

> Postfix (linguistics), an affix which is placed after the stem of a word

But when I click [the link](https://en.wikipedia.org/wiki/Postfix_(linguistics)), it redirects to [the article for "suffix".](https://en.wikipedia.org/wiki/Suffix)

Language is hard and dumb and arbitrary, and English more than most, and **_American_** English...

File: binaries/build_wheels.py
Comment (joeyparrish) [inline binaries/build_wheels.py:72]: So the `platform_binaries` argument is how you are avoiding using a temp directory?  You are allow-listing the binaries per-platform in each call to build a wheel?

**Review by joeyparrish — CHANGES_REQUESTED** [review]

**Comment by joeyparrish** [comment]
BTW, I really like the latest changes.  Much cleaner!

File: binaries/build_wheels.py
Comment (meryacine) [reply binaries/build_wheels.py:47]: yeah, i was just testing with your version because it has the right binary names.

File: binaries/build_wheels.py
Comment (meryacine) [reply binaries/build_wheels.py:72]: ummm... not exactly, 
platform_binaries is used to tell `setuptools` which files in the package(`streamer_binaries`) to include. By default, `setuptools` will include all the `.py` files and leave the others, will not put them into the wheel, by specifying them as `package_data`, we are telling `setuptools` to ship them into the `.whl` file it is going to generate with all the `.py` files.

The build directory problem is that when `setuptools` is building the package for one platform, it dumps every thing it will ship in `build/lib`(the default), then it doesn't delete it, the next platform in the for loop will be built in the same directory `build/lib`, but some old binaries are still there, thus we have to clean this directory after compressing the wheel for each platform.

We can't set `--build-lib` from the `bdist_wheel`. `bdist_wheel` calls `build`, and `build` has the `--build-lib` option, so i set the from the `build` command directly. By setting each build directory to the platform name we avoid reusing the same build directory so we are fine :D.

File: binaries/build_wheels.py
Comment (meryacine) [reply binaries/build_wheels.py:55]: I don't get it.  Do you mean i should remove all the indentation?

Comment (meryacine) [reply binaries/build_wheels.py:55]: got it, my bad.

**Review by joeyparrish — APPROVED** [review]
I still have questions about the mechanics of building the wheel, but I approve of what you've done regardless.  Let's find time to chat so I can learn more, and I may add some more comments based on our conversation.

**Comment by joeyparrish** [comment]
I spent some time playing around with the code, and I think I understand it now.  I may add a few comments later.

So as soon as you can resolve conflicts with this PR, I am ready to merge it.  Thank you for your hard work on this!  It's really exciting to see it landing.

**Comment by meryacine** [comment]
@joeyparrish While testing with `--test-hermetics` enabled, the `drmTests` for DASH and HLS fails.
```
+ /home/omar/.local/lib/python3.8/site-packages/streamer_binaries/packager-linux-x64 'in=/tmp/shaka-live-05_j15_r/5953b968-e71f-47fd-ba32-a663cc91e2b8,stream=video,init_segment=output_files/video_144p_108k_h264_init.mp4,segment_template=output_files/video_144p_108k_h264_$Number$.mp4' --quiet --segment_duration 4 --generate_static_live_mpd --mpd_output output_files/dash.mpd --hls_playlist_type VOD --hls_master_playlist_output output_files/hls.m3u8 --enable_widevine_encryption --key_server_url https://license.uat.widevine.com/cenc/getcontentkey/widevine_test --content_id EC71357493E0C07F58469229C8E31B38 --signer widevine_test --aes_signing_key 1ae8ccd0e7985cc0b6203a55855a1034afc252980e970ca90e5202689f947ab9 --aes_signing_iv d58ce954203b7c9a9a9d467f59839249 --protection_scheme cenc --clear_lead 0
[0904/092728:ERROR:proto_json_util.cc(34)] Failed to parse from JSON:  error: Unexpected end of string. Expected a value.

^
[0904/092728:ERROR:widevine_key_source.cc(409)] Failed to convert JSON to proto: 
[0904/092728:ERROR:packager_util.cc(80)] Widevine encryption key source failed to fetch keys: 13 (SERVER_ERROR): Failed to extract encryption key from ''.
Chrome 92.0.4515 (Linux 0.0.0) Shaka Streamer has widevine encryption enabled (hls) FAILED
        Error
            at <Jasmine>
            at UserContext.<anonymous> (tests.js:544:49)
        shaka.util.Error {
          "severity": 2,
          "category": 1,
          "code": 1001,
          "data": [
            "http://localhost:5000/output_files/hls.m3u8",
            500,
            "Some processes exited with non-zero exit codes",
            {
              "content-length": "46",
              "content-type": "text/html; charset=utf-8, text/plain"
            },
            0
          ],
          "handled": false
        } thrown
        Error: Failed to close Shaka Streamer
            at stopStreamer (tests.js:57:11)
            at async UserContext.<anonymous> (tests.js:80:5)
```
and 
```
+ /home/omar/.local/lib/python3.8/site-packages/streamer_binaries/packager-linux-x64 'in=/tmp/shaka-live-0xyurn9g/568c9820-33ed-4afd-a1c6-a998b03e169d,stream=video,init_segment=output_files/video_144p_108k_h264_init.mp4,segment_template=output_files/video_144p_108k_h264_$Number$.mp4' --quiet --segment_duration 4 --generate_static_live_mpd --mpd_output output_files/dash.mpd --hls_playlist_type VOD --hls_master_playlist_output output_files/hls.m3u8 --enable_widevine_encryption --key_server_url https://license.uat.widevine.com/cenc/getcontentkey/widevine_test --content_id EC71357493E0C07F58469229C8E31B38 --signer widevine_test --aes_signing_key 1ae8ccd0e7985cc0b6203a55855a1034afc252980e970ca90e5202689f947ab9 --aes_signing_iv d58ce954203b7c9a9a9d467f59839249 --protection_scheme cenc --clear_lead 0
[0904/092732:ERROR:proto_json_util.cc(34)] Failed to parse from JSON:  error: Unexpected end of string. Expected a value.

^
[0904/092732:ERROR:widevine_key_source.cc(409)] Failed to convert JSON to proto: 
[0904/092732:ERROR:packager_util.cc(80)] Widevine encryption key source failed to fetch keys: 13 (SERVER_ERROR): Failed to extract encryption key from ''.
Chrome 92.0.4515 (Linux 0.0.0) Shaka Streamer has widevine encryption enabled (dash) FAILED
        Error
            at <Jasmine>
            at UserContext.<anonymous> (tests.js:544:49)
        shaka.util.Error {
          "severity": 2,
          "category": 1,
          "code": 1001,
          "data": [
            "http://localhost:5000/output_files/dash.mpd",
            500,
            "Some processes exited with non-zero exit codes",
            {
              "content-length": "46",
              "content-type": "text/html; charset=utf-8, text/plain"
            },
            0
          ],
          "handled": false
        } thrown
        Error: Failed to close Shaka Streamer
            at stopStreamer (tests.js:57:11)
            at async UserContext.<anonymous> (tests.js:80:5)
```
I think it's a packager issue in the new static-build, it passes these tests with the old binaries i have in `PATH`.

**Comment by meryacine** [comment]
@joeyparrish Please don't merge this yet.

**Comment by joeyparrish** [comment]
> I think it's a packager issue in the new static-build, it passes these tests with the old binaries i have in PATH.

I found the issue (in Packager) and have a fix in review.  This will be followed by the official v2.6 release of Packager.

Thanks for catching this!

File: binaries/auto_build_setup.py
Comment (joeyparrish) [inline binaries/auto_build_setup.py:28]: I like this latest change overall, but it seems hacky to encode a python list into a single arg, then parse it using "ast", when we can naturally pass multiple args.  I would prefer we use a more typical command-line convention like "--" to separate the setuptools args from the list of binaries.

For example, the caller would look like:

```py
args = [
  'foo',
  'bar',
  '--py3',
  '--',
] + binary_paths
```

And here we would do something like:

```py
separator_index = sys.argv.index('--')
platform_binaries = sys.argv[separator_index+1:]
sys.argv = sys.argv[:separator_index]
```

What do you think?

File: binaries/setup.py
Comment (joeyparrish) [inline binaries/setup.py:25]: I think that this should be an error.  If there aren't any platform binaries, something is definitely wrong.

You could just let it raise the exception, or you could raise a RuntimeError with an explanatory message (like "don't run setup.py directly")

Comment (meryacine) [reply binaries/setup.py:25]: My intent was to have it usable as a normal setup.py.
But your point makes sense, it's all about the binaries.

**Review by joeyparrish — APPROVED** [review]
Looks good, but I will wait for your confirmation before I merge it.  Thanks!

**Comment by meryacine** [comment]
@joeyparrish I think it'a all fine now.  I have no more updates.
Thanks.

**Pull request merged**
